proposals/0001-lhls.md

```diff
@@ -0,0 +1,313 @@
+
+This specification provides no additional constraint on media type used. As of writing this, RFC8216 draft 03 supports MPEG-TS and fragmented MP4, including CMAF fmp4. Support for these formats is dependent on the client.
+
+## Glossary
+
+- Client: An MSE-Based JS player such as Hls.js
+- Server: The system comprising delivery, transcoding, cdn etc.; the party responsible for creating and serving the manifest
+- Prefetch segment: A segment which has been advertised but is not yet available
+- Complete segment: A segment which has been advertised and available. Any non-LHLS playlist consists only of complete segments.
+
+## Prefetch Media Segments
+
+The `EXT-X-PREFETCH` tag specifies a prefetch segment. Its format is:
+
+`#EXT-X-PREFETCH: <URI>`
+
+If delivering a low-latency stream, the server must deliver at least one prefetch segment, but no more than two. Doing so may introduce undue network burden. These segments must appear after all complete segments. Any constraint in the HLS spec applying to complete segments also applies to prefetch segments unless otherwise specified. For example:
+- A segment must remain available to clients for a period of time equal to the duration of the segment plus the duration of the longest Playlist file distributed by the server containing that segment.
+
+### Sequence Numbers
+[sequence-numbers]: #sequence-numbers
+
+A prefetch segment's Discontinuity Sequence Number is the value of the `EXT-X-DISCONTINUITY-SEQUENCE` tag (or zero if none) plus the number of `EXT-X-DISCONTINUITY` and `EXT-X-PREFETCH-DISCONTINUITY` tags in the Playlist preceding the URI line of the segment.
+
+If a prefetch segment is the first segment in a manifest, its Media Sequence Number is either 0, or declared in the Playlist. The Media Sequence Number of every other prefetch segment is equal to the Media Sequence Number of the complete segment or prefetch segment that precedes it plus one.
+
+### Playlist Duration
+
+The duration of a Media Playlist containing prefetch segments is considered to be equal to the duration of all complete segments plus the expected duration of all prefetch segments.
+
+## Media Segment Tags
+
+The server must not precede any prefetch segment with metadata other than those specified in this document, with the specified constraints.
+
+### EXTINF
+
+A prefetch segment must not be advertised with an `EXTINF` tag. The duration of a prefetch segment must be equal to or less than what is specified by the `EXT-X-TARGETDURATION` tag. 
+
+### EXT-X-DISCONTINUITY
+[discontinuities]: #ext-x-discontinuity
+
+A prefetch segment must not be advertised with an `EXT-X-DISCONTINUITY` tag. To insert a discontinuity just for prefetch segments, the server must insert the `EXT-X-PREFETCH-DISCONTINUITY` tag before the newest `EXT-X-PREFETCH` tag of the new discontinuous range. 
+
```

Comment:
In our use case, discontinuities are caused by a gap in timestamps in the stream sent by the broadcaster (because of a networking issue or a camera flip for instance). This means that we do not know if there is a discontinuity at the time we advertise the prefetch segment. We only know of the discontinuity at the point when we write the first data into the prefetch segment. 

This means that we would add a EXT-X-PREFETCH-DISCONTINUITY tag to a prefetch segment after it has first been advertised, and that a player may not see the EXT-X-PREFETCH-DISCONTINUITY tag by the time it starts playing back media from that prefetch segment.

This behavior doesn't seem to be explicitly forbidden by the spec, but it does require the player to have some heuristic to detect discontinuities that were not advertised in time.

Reply:
>This means that we would add a EXT-X-PREFETCH-DISCONTINUITY tag to a prefetch segment after it has first been advertised, and that a player may not see the EXT-X-PREFETCH-DISCONTINUITY tag by the time it starts playing back media from that prefetch segment.

I can foresee this being a problem if there is a large gap between the PTS values. Hls.js feeds a discontinuity tag into the muxer so that it can modify PTS values post-discontinuity to ensure that there is no gap. Without the tag we'd be inserting a gap equal to to the PTS time in seconds. Depending on how much forward buffer exists we can jump it without a stall, but we'd still need to do a seek in userland which would cause a momentary disruption.

Reply:
How would you imagine the interaction between EXT-X-GAP and prefetch segments should work? Should it be allowed to add EXT-X-GAP to an already advertised prefetch segment? That would be another way to address the use case above I believe.

Reply:
>This behavior doesn't seem to be explicitly forbidden by the spec

I'll add language which addresses this case. It doesn't make sense to add it as a prefetch discontinuity after the fact, but it must appear when transformed to a complete segment. Players should be able to handle gaps so I don't believe that this a problem. Maybe the prefetch discontinuity isn't even needed anymore - I need to think a bit harder about SSAI workflows. For SSAI you can probably just stop including prefetch segments and then insert a regular discontinuity.

Reply:
>How would you imagine the interaction between EXT-X-GAP and prefetch segments should work? Should it be allowed to add EXT-X-GAP to an already advertised prefetch segment? That would be another way to address the use case above I believe.

I suppose a gap would a) cause the client to ignore the segment or b) abort the connection if it has already been requested, but data has not yet been received. Not sure if I understand how to fix gaps with this - would you mind giving me an example?

In general the blocker for post-advertisement insertion of prefetch tags is that clients may not refresh the playlist in time to know that the segment has a new tag. I can see adding tags post-advertisement in the case that there are multiple prefetches and the tag is on N+1 or greater - the client should refresh the manifest before the encoder beings sending segments. But I think there will be some tricky race conditions with this kind of solution.

Reply:
@biglittlebigben in the case of a camera flip, how do you handle this now if it occurs in the middle of a segment?

I think that'll shed light onto how this particular case would interact.

Reply:
We currently detect the gap in timestamps in the backend. If it is over a threshold, we start a new segment and insert a discontinuity. This does mean that players watching over LHLS will not hear about the discontinuity in time from the playlist. Our player has heuristics to detect such gaps.

Reply:
So, I don't believe this makes much of a change to that workflow.

The prefetch segment up until the point the packager detects the gap in the backend would continue to push data normally like a prefetch. Upon detecting the gap, it would complete the segment, close the connection.

Updates the manifest locally with that completed segment, advertising the length now with a EXT-INF, push that manifest to the origin with the next prefetch having the DISCO applied to it.

Thoughts?

Reply:
That's what we would do indeed. Is this allowed by the spec though? (inserting a discontinuity for a segment that didn't have a prefetch discontinuity?)

Reply:
@biglittlebigben Yes, and I believe it should also be mandatory

Reply:
That would work for our use case then. Any feedback from player developers about having to detect and handle timestamp discontinuities?